server/src/test/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregatorTests.java

```diff
@@ -853,6 +913,225 @@ public void testNestedUnderTerms() throws IOException {
+                );
+
+                assertEquals(OUT_NESTED, nested.getName());
+                assertEquals(expectedNestedDocs, nested.getDocCount());
+
+                LongTerms outTerms = (LongTerms) nested.getProperty(OUT_TERMS);
+                assertEquals(1, outTerms.getBuckets().size());
+
+                InternalNested internalNested = (InternalNested) (((Object[]) outTerms.getProperty(INNER_NESTED))[0]);
+                assertEquals(expectedNestedDocs, internalNested.getDocCount());
+
+                LongTerms innerTerms = (LongTerms) internalNested.getProperty(INNER_TERMS);
+                assertEquals(bucketKeys.length, innerTerms.getBuckets().size());
+                for (int i = 0; i < expectedNestedDocs; i++) {
+                    LongTerms.Bucket bucket = innerTerms.getBuckets().get(i);
+                    assertEquals(bucketKeys[i], bucket.getKeyAsString());
+                    assertEquals(1, bucket.getDocCount());
+                }
+            }
+        }
+    }
+
+    private DocIdSetIterator getDocIdSetIterator(int[] value) {
+        int[] bits = new int[value[value.length - 1] + 1];
+        for (int i : value) {
+            bits[i] = 1;
+        }
+        return new DocIdSetIterator() {
+            int index = -1;
+
+            @Override
+            public int docID() {
+                if (index == -1 || index > bits.length || bits[index] != 1) {
+                    return -1;
+                }
+                return index;
+            }
+
+            @Override
+            public int nextDoc() {
+                for (int i = index; i < bits.length; i++) {
+                    if (bits[i] == 1) {
+                        index = i;
+                        return index;
+                    }
+                }
+                index = bits.length;
+                return -1;
+            }
+
+            @Override
+            public int advance(int target) {
+                for (int i = target; i < bits.length; i++) {
+                    if (bits[i] == 1) {
+                        index = i;
+                        return index;
+                    }
+                }
+                index = bits.length;
+                return -1;
+            }
+
+            @Override
+            public long cost() {
+                return bits.length;
+            }
+        };
+    }
+
+    public void testGetChildAndRootParent() throws IOException {
+        {
+            // p: parent c: child
+            // [p1], [c2,p3], [c4,p6] [p8]
+            BitSet parentDocs = new FixedBitSet(20);
+            parentDocs.set(0);
+            parentDocs.set(3);
+            parentDocs.set(6);
+            parentDocs.set(8);
+            DocIdSetIterator childDocs = getDocIdSetIterator(new int[] { 2, 4 });
+            Tuple<Integer, Integer> res = getChildAndParentId(parentDocs, childDocs, 0);
+            assertEquals(0, res.v1().intValue());
+            assertEquals(2, res.v2().intValue());
```

Comment:
This result looks wrong:

1. There is no doc with ID 1. Lucene doc ids do not have gaps.
2. The child document with ID 2 must have been indexed as a child of parent with ID 3. This return value suggests that the child with ID 2 is associated with the parent with ID 0.

Reply:
@msfroh Yes, you are right.
1. I will fix the logical after you finish reviewing the all code.
2. at the end of `getChildAndParentId`, we should add a check:  if currentParentDoc<=childDocId, childDocId = DocIdSetIterator.NO_MORE_DOCS.

please see other code.